<a href="https://colab.research.google.com/github/ParnianH98/Movie-Recommender-System/blob/main/Movie_Recommender_System.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
from google.colab import files
from google.colab import drive
drive.mount('/gdrive')
%cd /gdrive

Drive already mounted at /gdrive; to attempt to forcibly remount, call drive.mount("/gdrive", force_remount=True).
/gdrive


In [ ]:
df_cred = pd.read_csv(r'/gdrive/My Drive/AI HW/tmdb_5000_credits.csv')
df_movie = pd.read_csv(r'/gdrive/My Drive/AI HW/tmdb_5000_movies.csv')

In [ ]:
df_cred

movie_id  ...                                               crew
0        19995  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1          285  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2       206647  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3        49026  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4        49529  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...
...        ...  ...                                                ...
4798      9367  ...  [{"credit_id": "52fe44eec3a36847f80b280b", "de...
4799     72766  ...  [{"credit_id": "52fe487dc3a368484e0fb013", "de...
4800    231617  ...  [{"credit_id": "52fe4df3c3a36847f8275ecf", "de...
4801    126186  ...  [{"credit_id": "52fe4ad9c3a368484e16a36b", "de...
4802     25975  ...  [{"credit_id": "58ce021b9251415a390165d9", "de...

[4803 rows x 4 columns]

In [ ]:
df_movie.head(10)

budget  ... vote_count
0  237000000  ...      11800
1  300000000  ...       4500
2  245000000  ...       4466
3  250000000  ...       9106
4  260000000  ...       2124
5  258000000  ...       3576
6  260000000  ...       3330
7  280000000  ...       6767
8  250000000  ...       5293
9  250000000  ...       7004

[10 rows x 20 columns]

In [ ]:
df_cred.columns = ['id','tittle','cast','crew']
df = df_movie.merge(df_cred,on ='id')

# first approach

In [ ]:
filtered = df['vote_count'].quantile(0.8)
filtered

957.6000000000004

In [ ]:
q_movies = df.copy().loc[df['vote_count'] >= filtered]
q_movies.shape

(961, 23)

In [ ]:
meanVote = df['vote_average'].mean()
meanVote

6.092171559442011

In [ ]:
def weighted_rating(data, f = filtered, mv = meanVote):
    vc = data['vote_count']
    vav = data['vote_average']
    # Calculation based on the IMDB formula
    return (vc/(vc + f) * vav) + (f/(f + vc) * mv)

In [ ]:
# Define a new feature 'score' and calculate its value with `weighted_rating()`
q_movies['score'] = q_movies.apply(weighted_rating, axis=1)

In [ ]:
#Sort movies based on score calculated above
q_movies = q_movies.sort_values('score', ascending=False)

#Print the top 10 movies based on the score value
q_movies[['title', 'genres' , 'overview' , 'vote_count', 'vote_average', 'score']].head(10)


title  ...     score
1881  The Shawshank Redemption  ...  8.248353
662                 Fight Club  ...  8.096134
3337             The Godfather  ...  8.077404
3232              Pulp Fiction  ...  8.074738
65             The Dark Knight  ...  8.044250
809               Forrest Gump  ...  7.972814
96                   Inception  ...  7.969290
95                Interstellar  ...  7.937399
1990   The Empire Strikes Back  ...  7.904757
1818          Schindler's List  ...  7.900080

[10 rows x 6 columns]

# second approach

In [ ]:
#Import TfIdfVectorizer from scikit-learn
from sklearn.feature_extraction.text import TfidfVectorizer

#Define a TF-IDF Vectorizer Object. Remove all english stop words such as 'the', 'a'
tfiDF = TfidfVectorizer(stop_words='english')

#Overview is my choice
#Anybody else can use other facors to find similarities such as 'genre'
#Replace NaN with an empty string
df['overview'] = df['overview'].fillna('')

#Construct the required TF-IDF matrix by fitting and transforming the data
tfiDF_matrix = tfiDF.fit_transform(df['overview'])

#Output the shape of tfiDF_matrix
tfiDF_matrix.shape

(4803, 20978)

there are over 20000 words to descrie the 'overview' of 4803 films 

In [ ]:
# Import linear_kernel
from sklearn.metrics.pairwise import linear_kernel

# Compute the cosine similarity matrix using linear_kernel which is similar to cosine_similarity but this is faster
cosine = linear_kernel(tfiDF_matrix, tfiDF_matrix)

In [ ]:
#Construct a reverse map of indices and movie titles
indices = pd.Series(df.index, index=df['title']).drop_duplicates()

In [ ]:
# Function that takes in movie title as input and outputs most similar movies
def get_recommendations(title, cosine=cosine):
    # Get the index of the movie that matches the title
    #step 1
    idx = indices[title]

    # Get the pairwsie similarity scores of all movies with that movie
    #step 2
    sim_scores = list(enumerate(cosine[idx]))

    # Sort the movies based on the similarity scores
    #step 3
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar movies
    # step 4
    sim_scores = sim_scores[1:11]

    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]

    # Return the top 10 most similar movies
    return df[['title', 'genres' , 'overview' , 'vote_count', 'vote_average']].iloc[movie_indices]

In [ ]:
get_recommendations('Inception')

title  ... vote_average
2897                              Cypher  ...          6.7
134   Mission: Impossible - Rogue Nation  ...          7.1
1930                          Stone Cold  ...          5.7
914                 Central Intelligence  ...          6.2
1683                     Pitch Perfect 2  ...          6.8
1248                      At First Sight  ...          5.9
1512               A History of Violence  ...          6.9
2389                         Renaissance  ...          6.7
1803                      Blood and Wine  ...          5.5
1267                              Duplex  ...          5.9

[10 rows x 5 columns]